### ISSUE TITLE
How to choose decay rate? (No success with RAdam - does one need a decay scheduler or gradient clipping)

### ISSUE THREAD

**Issue by brando90** [LiyuanLucasLiu/RAdam#66]
Hi Liyuan,

I've had difficulties with RAdam in my sequence learning problems. I am using a standard pytorch transformer with your library. I was wondering if one usually needs a scheduler (e.g. to reduce the learning rate at the end) besides RAdam or gradient clipping since my models seem to diverge often:

![Screen Shot 2021-08-05 at 10 37 03 AM](https://user-images.githubusercontent.com/1855278/128387970-0eae58b6-d86b-4dd9-b2e6-f4d06c17ca30.png)

**Comment by brando90** [comment]
I also get this:

![Screen Shot 2021-08-05 at 11 45 13 AM](https://user-images.githubusercontent.com/1855278/128388911-0f3bccb9-63b7-4aac-b00c-d5e0d5746a3d.png)

**Comment by brando90** [comment]
note most of these are small data sets (500-1000). Not sure if that matters. First one with batch size 128 (data set size 1000) second one with batch size equal to the size of the data set 500.

**Comment by brando90** [comment]
At the core I think RAdam seems to need annealing or decay rate. Does RAdam have recommendations?

**Comment by LiyuanLucasLiu** [comment]
Thanks for reaching out!

Let me first give you a quick answer:

1. whether we need a learning rate scheduler? I believe in most cases, learning rate scheduler is needed. From my experience, I usually prefer linear decay with warmup or cosine annealing with restart. I would suggest you to start from trying these. 

2. how to stabilize Transformer. Transformer has been found to be harder to train. You may want to check our [Admin](https://github.com/LiyuanLucasLiu/Transformer-Clinic) project. 

3. whether we still need gradient clipping. I believe it is usually helpful to apply this to stabilize the training. 

4. if there are some prior work with Transformer having a similar setting with yours, maybe try their setup first.

**Comment by brando90** [comment]
Thanks for the response Liyuan! I appreciate the response! (maybe some time I get back to illinois we could meet in person after covid is cover...)

I will check those suggestions. 

Out of curiosity, do you have any experience with AdaFactor (https://huggingface.co/transformers/main_classes/optimizer_schedules.html) and if yes what do you think of it? How well does it compare to the techniques you've developed (e.g. admin, radam, others)?

Thanks again!
